In [7]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report

In [8]:
# 1. Load data
df = pd.read_csv('../Data/marketing_campaign.csv', sep=';')
df.head()

ID  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0  5524        1957  Graduation         Single  58138.0        0         0   
1  2174        1954  Graduation         Single  46344.0        1         1   
2  4141        1965  Graduation       Together  71613.0        0         0   
3  6182        1984  Graduation       Together  26646.0        1         0   
4  5324        1981         PhD        Married  58293.0        1         0   

  Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  AcceptedCmp3  \
0  2012-09-04       58       635  ...                  7             0   
1  2014-03-08       38        11  ...                  5             0   
2  2013-08-21       26       426  ...                  4             0   
3  2014-02-10       26        11  ...                  6             0   
4  2014-01-19       94       173  ...                  5             0   

   AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  Complain  \
0             0             0             0             0         0   
1             0             0             0             0         0   
2             0             0             0             0         0   
3             0             0             0             0         0   
4             0             0             0             0         0   

   Z_CostContact  Z_Revenue  Response  
0              3         11         1  
1              3         11         0  
2              3         11         0  
3              3         11         0  
4              3         11         0  

[5 rows x 29 columns]

In [9]:
# 2. Preprocess
# Handle missing, outliers, encode categorical (Education, Marital) if needed
df['Response'] = df['Response'].astype(int)  # ensure it's 0/1

In [10]:
# 3. Select features & target
features = [
    'AcceptedCmp1', 'AcceptedCmp2', 'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5',
    'Complain', 'Income', 'MntWines', 'MntMeatProducts', 'NumWebPurchases',
    'NumStorePurchases', 'Recency', 'NumDealsPurchases'
]
X = df[features]
y = df['Response']  # target

In [11]:
# 4. Split data
X_train, X_test, y_train, y_test = train_test_split(X, y,
                                                    test_size=0.3,
                                                    random_state=42)

In [12]:
# 5. Train model
model = RandomForestClassifier(random_state=42)
model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [13]:
# 6. Evaluate
y_pred = model.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.90      0.97      0.93       577
           1       0.61      0.33      0.42        95

    accuracy                           0.88       672
   macro avg       0.75      0.65      0.68       672
weighted avg       0.86      0.88      0.86       672



In [14]:
# 7. Check feature importances
importances = model.feature_importances_
for feat, imp in zip(features, importances):
    print(f"{feat}: {imp:.4f}")

AcceptedCmp1: 0.0603
AcceptedCmp2: 0.0094
AcceptedCmp3: 0.0565
AcceptedCmp4: 0.0160
AcceptedCmp5: 0.0617
Complain: 0.0013
Income: 0.1448
MntWines: 0.1428
MntMeatProducts: 0.1397
NumWebPurchases: 0.0676
NumStorePurchases: 0.0819
Recency: 0.1615
NumDealsPurchases: 0.0566


Below is an interpretation of the feature importances from your model, which predicts a customer’s likelihood to accept the latest marketing campaign (Response). Each feature’s importance indicates how much it contributes to the model’s predictive power.

Recency: 0.1615
Income: 0.1448
MntWines: 0.1428
MntMeatProducts: 0.1397
NumStorePurchases: 0.0819
NumWebPurchases: 0.0676
AcceptedCmp5: 0.0617
AcceptedCmp1: 0.0603
NumDealsPurchases: 0.0566
AcceptedCmp3: 0.0565
AcceptedCmp4: 0.0160
AcceptedCmp2: 0.0094
Complain: 0.0013

1. Top Predictors
	1.	Recency (0.1615)
	•	Meaning: The number of days since the last purchase is the strongest indicator of whether someone will accept the new offer.
	•	Insight: Customers who purchased more recently tend to be more engaged and more likely to respond to campaigns.
	•	Action: Prioritize outreach to those who’ve purchased recently—e.g., schedule follow-up offers soon after a purchase.
	2.	Income (0.1448)
	•	Meaning: Higher or lower income directly influences the propensity to respond.
	•	Insight: Customers with higher disposable income might be more receptive to offers (particularly premium products).
	•	Action: Tailor campaign messages or product offerings by income segment. For high-income segments, highlight premium or luxury products; for lower-income, focus on deals or value propositions.
	3.	MntWines (0.1428) & MntMeatProducts (0.1397)
	•	Meaning: Amount spent on wines and meats are both substantial predictors.
	•	Insight: Customers who spend more on these categories may have specific purchasing preferences or lifestyles.
	•	Action: Segment or personalize offers around these product categories. For example, wine enthusiasts might respond well to exclusive wine promotions.

2. Medium-Level Predictors
	4.	NumStorePurchases (0.0819) & NumWebPurchases (0.0676)
	•	Meaning: The count of in-store vs. online purchases.
	•	Insight: Customers’ channel preference (store vs. web) matters for campaign response. Those with higher in-store purchases might prefer physical touchpoints or in-store coupons; heavy web shoppers might respond to online deals or email campaigns.
	•	Action: Use these channel insights to deliver the right offer in the preferred channel—e.g., send store-related promotions to in-store buyers.
	5.	AcceptedCmp5 (0.0617), AcceptedCmp1 (0.0603), AcceptedCmp3 (0.0565)
	•	Meaning: Past acceptance of campaigns (especially campaign #5, #1, and #3) is correlated with responding to the latest offer.
	•	Insight: Customers who accepted previous offers are more likely to respond again.
	•	Action: Re-target loyal responders. If a customer has a history of saying “yes,” they’re prime candidates for future offers.
	6.	NumDealsPurchases (0.0566)
	•	Meaning: Number of discounted purchases indicates price-sensitivity or deal-seeking behavior.
	•	Action: Highlight discounts or special offers to these customers to boost acceptance rates.

3. Lower-Level Predictors
	7.	AcceptedCmp4 (0.0160), AcceptedCmp2 (0.0094)
	•	Meaning: History of accepting the 4th and 2nd campaigns still matters, but less so than acceptance of other campaigns.
	•	Insight: There’s some correlation between these older campaigns and future acceptance, but not as strong as campaigns #1, #3, or #5.
	8.	Complain (0.0013)
	•	Meaning: Whether a customer complained in the last 2 years has minimal impact on whether they accept a new offer.
	•	Insight: Complaining behavior might indicate dissatisfaction, but it doesn’t necessarily predict campaign acceptance.
	•	Action: Complaints may be more important for customer service or churn analysis than for short-term campaign acceptance.

4. Key Takeaways & Actions
	1.	Recency is King
	•	Customers who purchased recently are most responsive. Focus on timely follow-ups—e.g., a new product offer sent within weeks of their last purchase.
	2.	Income & Category Spending
	•	High spenders (especially on wines and meats) appear more likely to accept new offers—possibly due to lifestyle or affordability factors.
	•	Tailor offers to match their spending patterns (e.g., premium or gourmet items for big spenders).
	3.	Channel Preferences
	•	Look at NumStorePurchases vs. NumWebPurchases to personalize marketing channels. In-store buyers may appreciate physical mail or in-store events, while web buyers respond well to email or website promotions.
	4.	History of Accepting Offers
	•	Prior campaign acceptance signals a “warm audience.” These customers often convert again, so re-target them.
	•	Specifically note that campaigns #1, #3, and #5 acceptance are stronger signals than #2 or #4.
	5.	Complaint is Not a Deal Breaker
	•	Complaints have very little impact on acceptance. They might be more relevant for customer satisfaction or churn, but here, they don’t strongly predict response.

Overall Strategy
	•	Time-Based Targeting: Prioritize customers with low recency (recent engagement).
	•	Segment by Spending Profile: Income, MntWines, and MntMeatProducts are crucial for personalizing offers.
	•	Channel Optimization: Distinguish store-based vs. web-based preferences.
	•	Leverage Loyal Responders: People who accepted earlier campaigns remain a hot lead.

Using these insights, you can create more effective, data-driven campaigns that reach the right customers with the right offers at the right time—ultimately increasing response rates and improving ROI.